src/core/server/saved_objects/saved_objects_service.ts

```diff
@@ -0,0 +1,116 @@
+import { retryCallCluster } from '../elasticsearch/retry_call_cluster';
+import { Logger } from '..';
+
+/**
+ * @public
+ */
+// eslint-disable-next-line @typescript-eslint/no-empty-interface
+export interface SavedObjectsServiceSetup {}
+
+/**
+ * @public
+ */
+export interface SavedObjectsServiceStart {
+  migrator: Pick<KibanaMigrator, keyof KibanaMigrator>;
+}
+
+/** @internal */
+export interface SavedObjectsSetupDeps {
+  legacy: LegacyServiceSetup;
+  elasticsearch: ElasticsearchServiceSetup;
+}
+
+/** @internal */
+// eslint-disable-next-line @typescript-eslint/no-empty-interface
+export interface SavedObjectsStartDeps {}
+
+export class SavedObjectsService
+  implements CoreService<SavedObjectsServiceSetup, SavedObjectsServiceStart> {
+  private migrator: KibanaMigrator | undefined;
+  logger: Logger;
+
+  constructor(private readonly coreContext: CoreContext) {
+    this.logger = coreContext.logger.get('savedobjects-service');
+  }
+
+  public async setup(coreSetup: SavedObjectsSetupDeps) {
+    this.logger.debug('Setting up SavedObjects service');
+
+    const {
+      savedObjectSchemas,
+      savedObjectMappings,
+      savedObjectMigrations,
+      savedObjectValidations,
+    } = await coreSetup.legacy.uiExports;
+
+    const adminClient = await coreSetup.elasticsearch.adminClient$.pipe(first()).toPromise();
+
+    const kibanaConfig = await this.coreContext.configService
+      .atPath<KibanaConfig>('kibana')
+      .pipe(first())
+      .toPromise();
+
+    this.migrator = new KibanaMigrator({
+      savedObjectSchemas,
+      savedObjectMappings,
+      savedObjectMigrations,
+      savedObjectValidations,
+      logger: this.coreContext.logger.get('migrations'),
+      kibanaVersion: this.coreContext.env.packageInfo.version,
+      config: coreSetup.legacy.pluginExtendedConfig,
+      kibanaConfig: (kibanaConfig as any) as KibanaConfig,
+      callCluster: retryCallCluster(adminClient.callAsInternalUser.bind(adminClient)),
+    });
+
+    return ({} as any) as Promise<SavedObjectsServiceSetup>;
+  }
+
+  public async start(core: SavedObjectsStartDeps): Promise<SavedObjectsServiceStart> {
+    this.logger.debug('Starting SavedObjects service');
+
+    /**
+     * Note: We want to ensure that migrations have completed before
+     * continuing with further Core startup steps that might use SavedObjects
+     * such as running the legacy server, legacy plugins and allowing incoming
+     * HTTP requests.
+     *
+     * However, our build system optimize step and some tests depend on the
+     * HTTP server running without an Elasticsearch server being available.
+     * So, when the `--skip-migrations` flag is true, we skip migrations all-
+     * together.
+     */
+    const cliArgs = this.coreContext.env.cliArgs;
+    const skipMigrations = cliArgs.optimize || cliArgs.skipMigrations;
+    await this.migrator!.awaitMigration(skipMigrations);
+
+    return { migrator: this.migrator as KibanaMigrator };
```

Comment:
why not define migrator as setup contract? Legacy has access to both setup and start contract https://github.com/restrry/kibana/blob/6e1a527d4a8e4bf9beb8e0e8dd0e7eb966f00843/src/legacy/server/kbn_server.js#L65-L66

Reply:
This is a little bit theoretical since it's not really possible for any Plugins to call this API, but my thinking was that the migrator allows you to start migrations with `awaitMigrations()`. I didn't want to expose that functionality as we want to be sure migrations only run once all setup is complete. So the SavedObjectsService "hides" the migrator until after we've run migrations.

In the future the SavedObjectsService will expose a migration API to `CoreSetup` that allows Plugins to register new migrations, but even then I don't think anyone should be able to access the migrator instance itself.